src/Microsoft.CodeAnalysis.Razor/RequiredAttributeParser.cs

```diff
@@ -0,0 +1,304 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Evolution;
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    // Internal for testing
+    internal static class RequiredAttributeParser
```

Comment:
That'd put the logic of CSS based required attributes in the builder and not in the `DefaultTagHelperDescriptorFactory`. That sounds a little backwards compared to how all other declarative information on TagHelper's is consumed.